std/typecons.d

```diff
@@ -7838,3 +7838,223 @@ unittest
     assert(~Ternary.no == Ternary.yes);
     assert(~Ternary.unknown == Ternary.unknown);
 }
+
+/**
+Provides untyped storage of size `Size` aligned to `Alignment`.
+
+`Alignment` must be a power of 2.
+Default `Alignment` is the maximum alignment requirement on the current platform.
+
+Note: while this type ensures storage alignment on its own, it still obeys
+language rules. Therefore, care must be taken when using it in aggregate types,
+notably with custom `align` specification.
+
+Since aggregated object cannot test it's alignment in an aggregate
+at compile time, an assert will be issued at runtime when a misaligned
+AlignedStorage is used.
+
+Example:
+
+    struct A
+    {
+        bool flag;
+        AlignedStorage!(4,4) storage; // OK, storage will be aligned to 4 bytes
+    }
+
+    struct B
+    {
+        align(1):
+        bool flag;
+        AlignedStorage!(4,4) storage; // WARNING: storage will be aligned to 1 byte
+    }
+
+    A a;
+    B b;
+    import std.conv : emplace;
+    emplace(cast(int*)(a.storage.ptr), 42); // OK
+    emplace(cast(int*)(b.storage.ptr), 42); // an AssertError will be thrown
+*/
+struct AlignedStorage(size_t Size, size_t Alignment = platformAlignment)
+{
+    /// Size of storage, in bytes
+    enum size = Size;
+
+    /// Alignment of storage, in bytes
+    enum alignment = Alignment;
+
+    @system @nogc nothrow:
+
+    /// Returns a pointer to the beginning of storage
+    @property inout(void)* ptr() inout
+    {
+        return &storage[0];
+    }
+
+    /// Returns storage as a slice
+    inout(void[]) opSlice() inout
+    {
+        return storage[];
+    }
+
+    /// Returns a slice `i`..`j` of storage
+    inout(void[]) opSlice(size_t i, size_t j) inout
+    {
+        return storage[i..j];
+    }
+
+    /// Returns size of storage
+    size_t opDollar() const
+    {
+        return size;
+    }
+
+    @trusted invariant()
+    {
+        // can't test this at compile time
+        assert(((cast(size_t)storage.ptr) % alignment) == 0,
+                            "AlignedStorage alignment violation");
+    }
+
+private:
+    align(Alignment)
+    void[Size] storage;
+}
+
+unittest
+{
+    import std.range : iota;
+    import core.bitop : bsr;
+    // test all alignments up to and including platformAlignment
+    foreach(a; aliasSeqOf!(iota(bsr(platformAlignment)+1)))
+        static assert(AlignedStorage!(1,2^^a).alignof == 2^^a);
+}
+
+unittest
+{
+    import std.range : iota;
+    import core.bitop : bsr;
+    import std.format : format;
+    // test all alignments up to and including platformAlignment
+    foreach(a; aliasSeqOf!(iota(bsr(platformAlignment)+1)))
+    {
+        mixin(q{
+            static struct S%1$d
+            {
+                bool b;
+                AlignedStorage!(1,2^^%1$d) s;
+            }
+
+            static assert(S%1$d.s.offsetof == 2^^%1$d);
+        }.format(a));
+    }
+}
+
+
+// catch misalgnment at runtime
+debug unittest
```

Comment:
Note that `debug` blocks (activated by `-debug` switch) may not necessarily be active at the same time as asserts and invariants (deactivated by `-release` switch).